File: android/src/main/java/au/com/crazybean/mobilex/kurir/modules/tasks/TasksDelegate.kt
package au.com.crazybean.mobilex.kurir.modules.tasks

import android.os.Bundle
import androidx.lifecycle.Observer
import au.com.crazybean.foundation.mvvm.Delegate
import au.com.crazybean.mobilex.kurir.data.model.Task

class TasksDelegate(view: TasksView?, viewModel: TasksViewModel) : Delegate<TasksView, TasksViewModel>(view, viewModel) {

    override fun onCreate(params: Bundle) {
        super.onCreate(params)
        onRefresh()
    }

    fun onRefresh() {
        view?.showSpinner()
        viewModel.tasks
            .observe(this, Observer { tasks ->
                view?.hideSpinner()
                tasks?.takeIf { it.isNotEmpty() }?.let {
                    view?.showTasks(it)
                }?: view?.showEmpty()
            })
    }

    fun onNewTask() {
        view?.showCreation()
    }

    fun onViewDetail(task: Task) {
    }
}

File: ios/Kurir/modules/tasks/TasksDelegate.swift
//
//  FindDelegate.swift
//  Kurir
//
//  Created by Loren on 19/9/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Crazybean

class TasksDelegate: Delegate<TasksView, TasksViewModel> {
    override func onCreate(_ params: [String : Any?]? = nil) {
        super.onCreate(params)
        onRefresh()
    }
    
    private func onRefresh() {
        view?.showSpinner()
        viewModel.loadTasks().observe(for: self) { [weak self] tasks in
            self?.view?.hideSpinner()
            guard let tasks = tasks else {
                self?.view?.showEmpty()
                return
            }
            
            self?.view?.showTasks(tasks: tasks)
        }
    }
}
